pkg/controller/core.oam.dev/v1alpha2/application/revision.go

```diff
@@ -0,0 +1,156 @@
+		Spec: v1alpha2.ApplicationRevisionSpec{
+			Application:              copiedApp,
+			ComponentDefinitions:     nil, // TODO: add componentDefinition when completed.
+			Components:               comps,
+			ApplicationConfiguration: ac,
+		},
+	}
+	appRev.Name = appfile.RevisionName
+	appRev.Namespace = application.Namespace
+
+	for _, w := range appfile.Workloads {
+		if w == nil {
+			continue
+		}
+		if !DupWorkloadDef(appRev, w.FullTemplate.WorkloadDefinition) {
+			// TODO(wonderflow): the workloadDefinition is going to refactor to componentDefinition, so we record it as componentDefinition here
+			hashValue, err := utils.ComputeSpecHash(w.FullTemplate.WorkloadDefinition.Spec.DeepCopy())
+			if err != nil {
+				return false, nil, errors.Wrapf(err, "fail to compute hash for workloadDefinition %s", w.FullTemplate.WorkloadDefinition.Name)
+			}
+			componentDefinitionHash = append(componentDefinitionHash, v1alpha2.NameAndHashValue{
+				Name:      w.FullTemplate.WorkloadDefinition.Name,
+				HashValue: hashValue,
+			})
+			appRev.Spec.WorkloadDefinitions = append(appRev.Spec.WorkloadDefinitions, w.FullTemplate.WorkloadDefinition)
+		}
+		for _, t := range w.Traits {
+			if t == nil {
+				continue
+			}
+			if !DupTraitDef(appRev, t.FullTemplate.TraitDefinition) {
+				hashValue, err := utils.ComputeSpecHash(t.FullTemplate.TraitDefinition.Spec.DeepCopy())
+				if err != nil {
+					return false, nil, errors.Wrapf(err, "fail to compute hash for traitDefinition %s", t.FullTemplate.TraitDefinition.Name)
+				}
+				traitDefinitionHash = append(traitDefinitionHash, v1alpha2.NameAndHashValue{
+					Name:      t.FullTemplate.TraitDefinition.Name,
+					HashValue: hashValue,
+				})
+				appRev.Spec.TraitDefinitions = append(appRev.Spec.TraitDefinitions, t.FullTemplate.TraitDefinition)
+			}
+		}
+	}
+	newRevision := &v1alpha2.AppRevisionHash{
+		ApplicationSpec:      appHash,
+		ComponentDefinitions: componentDefinitionHash,
+		TraitDefinitions:     traitDefinitionHash,
+	}
+	// check API spec difference
+	if application.Status.LatestRevision != nil && application.Status.LatestRevision.AppRevisionHash != nil {
+		// get the last revision and double check
+		lastAppRevision := &v1alpha2.ApplicationRevision{}
+		if err := r.Get(ctx, client.ObjectKey{Name: application.Status.LatestRevision.Name, Namespace: application.Namespace}, lastAppRevision); err != nil {
+			return false, nil, errors.Wrapf(err, "fail to get applicationRevision %s", application.Status.LatestRevision.Name)
+		}
+		if DeepEqualRevision(lastAppRevision, appRev) {
+			// No difference on spec, will not create a new revision
+			return false, nil, nil
+		}
+	}
+
+	return true, newRevision, r.Create(ctx, appRev)
+}
+
+// DeepEqualRevision will check the Application and Definition to see if the Application is the same revision
+// We don't check the rendered AC and Component spec here, because their spec type is runtime.RawExtension which is hard to check.
```

Comment:
All the definitions contain runtime.RawExtension, why can we compare there but not in the AC? I think it's actually enough to just compare AC if it refers to component revision